### PR TITLE
Enumerable public methods definition

### PR DESCRIPTION
**Pull request by ricardovaltierra**
The following methods where defined inside of 'Enumerable' module:
- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

Each of them with the purpose of being the own version of Ruby's Enumerable module
Also was defined 'multiply_els' method for testing purposes on 'my_inject'

* Every method was tested against the code on 'test.rb' file, wich due to microverse requirements was added to .gitignore

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+module Enumerable
+  def my_each
+    temp = self
+    i = 0
+    while i < temp.length
+      yield(temp[i])
+      i += 1
+    end
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_each`
- enumerables my_each returns an Enumerator if no block is given

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_each_with_index
+    temp = self
+    i = 0
+    while i < temp.length
+      yield(temp[i], i)
+      i += 1
+    end
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_each_with_index`
- enumerables my_each_with_index returns an enumerator if no block is given

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_select
+    temp = []
+    my_each { |e| temp << e if yield(e) }
+    temp
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_select`
- enumerables my_select returns an enumerator if no block is given

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_all?
+    test = true
+    my_each { |d| break unless test == yield(d) }
+    test
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_all?`
- enumerables my_all? returns true if the block never returns false or nil
rspec ./spec/spec_enumerables_spec.rb:69 # enumerables my_all? when no block or argument is given returns true when none of the collection members are false or nil
- enumerables my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- enumerables my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- enumerables my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_any?
+    test = false
+    my_each { |u| break if test == yield(u) }
+    test
+  end
+
```
Comment (imhta) [inline enumerable.rb]: In `my_any?`
- enumerables my_any? returns true if the block ever returns a value other than false or nil
- enumerables my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- enumerables my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- enumerables my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- enumerables my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  def my_none?
+    test = true
+    my_each { |t| break unless test == !yield(t) }
+    test
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_none?`
- enumerables my_none? returns true if the block never returns true for all elements
- enumerables my_none? when no block or argument is given returns true only if none of the collection members is true
- enumerables my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- enumerables my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- enumerables my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_count
+    total = 0
+    my_each { |z| block_given? ? (total += 1 if yield(z)) : (total += 1) }
+    total
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_count`
- enumerables my_count counts the number of items in enum that are equal to item if an argument is given

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+  def my_map(&proc)
+    temp = []
+    my_each { |f| defined?(proc) ? (temp << proc.call(f)) : (temp << yield(f)) }
+    temp
+  end
```
Comment (imhta) [inline enumerable.rb]: In `my_map`
- enumerables my_map returns an Enumerator if no block is given

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    if !init.nil?
+      memo = init
+      my_each_with_index { |v| memo = yield(memo, v) }
+    else
+      memo = self[0]
+      my_each_with_index { |v, i| memo = yield(memo, v) if i.positive? }
+    end
+    memo
+  end
+end
```
Comment (imhta) [inline enumerable.rb]: In `my_inject`
- enumerables  my_inject combines all elements of enum by applying a binary operation, specified by a block
- enumerables my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by imhta — CHANGES_REQUESTED** [review]
# Changes required 🛑 

### Hello @ricardovaltierra , 👋 there some tests are failing, for more details check inline comments. and fix them.

---
#### Don't forgot to request for another code review
---

**Pon Muthu Selvam N**
_If you have any queries, feel free to ping me on slack_

**Comment by ricardovaltierra** [comment]
Changes made:

Refactoring of thenext methods:

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

Thanks :)

**Comment by shshamim63** [comment]
## Status :recycle: [!! Required Changes !!] :recycle: 
___
Hi @ricardovaltierra :wave: ,
massive update :wave: . You have covered most of the task mentioned my the previous reviewer. But still there are some fields that you need to cover to complete this project. And I am sure you will cover all of them.
___
#### Please follow the instructions below :point_down: to complete this project.
___
### :muscle: [Mandatory Changes] :muscle: 
___
1. **my_each** -> Returns an Enumerator if no block is given
image.
For an example I have modified your my_each 
https://github.com/ricardovaltierra/enumerable-methods/blob/4a2d9fc2d38697928e44f56767630ef159576629/enumerable.rb#L5-L14
in this way :point_down: 
```
def my_each
    return to_enum(:enum) unless block_given?
    temp = self
    i = 0
    while i < temp.length
      yield(temp[i])
      i += 1
    end
    temp
end
```
2. **my_each_with_index**-> Returns an enumerator if no block is given
Check this [link](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). The way to cover it is in :point_up: .
3. **my_select**->Returns an enumerator if no block is given. Also same as above.
4. **my_all** -> Returns true if all of the collection is a member of such class. Check this [link](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)
For an example, `[1, 2, 3].my_all?(Integer)` should return true.
5. **my_map** ->Returns an enumerator if no block is given. Also same as above.
___
### :clubs: [Optional Changes] :clubs: 
___
I have tried to refactor your my_all? and came up with this. If you think this is way more short than your one, you can apply the way i have used.
```
def my_all?(pattern = nil)
    if pattern # If a pattern is given
      my_each { |element| return false unless pattern === element }
    elsif block_given? # If a block is given
      my_each { |element|  return false unless yield(element) }
    else # If nothing is given
      my_each { |element| return false unless element }
    end
    true
 end
```
#### Please fix the above mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](Shakhawat Hossain)

**Comment by ricardovaltierra** [comment]
The following modifications were made:

- my_each -> Returns an Enumerator if no block is given
- my_each_with_index -> Returns an enumerator if no block is given
- my_select -> Returns an enumerator if no block is given
- my_all -> Returns true if all of the collection is a member of such class. * Test made with `[1, 2, 3].my_all?(Integer)`
- my_map -> Returns an enumerator if no block is given.
- my_all? -> Refactored with recomended code to improve performance

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      if block_given?
+        temp << yield(f)
+      elsif temp << defined?(proc)
+        proc.call(f)
+      else
+        return to_enum(:enum)
+      end
+    end
+    temp
+  end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]**
* Returns an Enumerator if no block is given

Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]**
* Returns an Enumerator if no block is given

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required 
### Great work implementing the changes suggested by the other TSE(s) :clap: 
We only have two more issues to address before moving on. See the comments :point_down: 

**Happy Coding :computer: 
Simon Wathigo**

**Comment by ricardovaltierra** [comment]
The following changes were made:

-Returns an Enumerator if no block is given (line 85)

**Pull request merged**